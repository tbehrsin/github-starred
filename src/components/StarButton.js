import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { Star } from 'styled-icons/octicons';

const StarButton = styled.button`
  font: inherit;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  margin-left: 8px;
  font-size: 12px;
  line-height: 20px;
  padding: 3px 10px;
  background: #eff3f6 linear-gradient(-180deg,#fafbfc,#eff3f6 90%);
  color: #24292e;
  border: 1px solid rgba(27, 31, 35, 0.2);
  border-radius: 0.25em;
  cursor: pointer;
  font-weight: bold;

  & > * + * {
    margin-left: 0.4em;
  }

  &:hover {
    background: #e6ebf1 linear-gradient(-180deg,#f0f3f6,#e6ebf1 90%);
    border: 1px solid rgba(27, 31, 35, 0.35);
  }

  &:active {
    background: #e9ecef;
    border: 1px solid rgba(27, 31, 35, 0.35);
    box-shadow: inset 0 0.15em 0.3em rgba(27, 31, 35, 0.15);
  }
`;

const ADD_STAR = gql`
  mutation AddStar($id: ID!) {
    addStar(input: { starrableId: $id }) {
      clientMutationId
      starrable {
        viewerHasStarred
      }
    }
  }
`;

const REMOVE_STAR = gql`
  mutation RemoveStar($id: ID!) {
    removeStar(input: { starrableId: $id }) {
      clientMutationId
      starrable {
        viewerHasStarred
      }
    }
  }
`;

export default ({ repository, onChange }) => (
  <Mutation mutation={repository.viewerHasStarred ? REMOVE_STAR : ADD_STAR}>
    {(star, { data }) => {
      const starred = data ?
        (data.addStar || data.removeStar).starrable.viewerHasStarred :
        repository.viewerHasStarred;

      if (data) {
        onChange(starred);
      }

      if (starred) {
        return (
          <StarButton onClick={() => star({ variables: { id: repository.id }})}>
            <Star size={16} />
            <span>Unstar</span>
          </StarButton>
        );
      }

      return (
        <StarButton onClick={() => star({ variables: { id: repository.id }})}>
          <Star size={16} />
          <span>Star</span>
        </StarButton>
      );
    }}
  </Mutation>
);
