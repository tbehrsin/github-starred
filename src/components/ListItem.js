import React from 'react';
import styled from 'styled-components';
import { Repo, Star } from 'styled-icons/octicons';

import StarButton from './StarButton';

const Container = styled.div`
  border: 1px solid #d1d5da;
  border-radius: 3px;
  background: white;
  padding: 16px;
  box-sizing: border-box;
  margin: 16px;

  & > * + * {
    margin-top: 8px;
  }
`;

const Row = styled.div`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
`;

const RepositoryName = styled.a`
  font-weight: bold;
  text-decoration: none;
  color: inherit;

  & > * + * {
    margin-left: 0.3em;
  }
`;

const Description = styled.div`
  font-size: 15px;
  line-height: 22px;
  color: #586069;
`;

const Stargazers = styled.div`
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  color: #586069;
  font-size: 12px;

  & > * + * {
    margin-left: 0.3em;
  }
`;



export default ({ repository, onChange }) => {
  return (
    <Container>
      <Row>
        <RepositoryName href={repository.url} target="_new">
          <Repo size={16} />
          <span>{repository.owner.login}/{repository.name}</span>
        </RepositoryName>
        <StarButton repository={repository} onChange={onChange} />
      </Row>
      <Description dangerouslySetInnerHTML={{ __html: repository.descriptionHTML }} />
      <Stargazers>
        <Star size={12} />
        <span>{repository.stargazers.totalCount}</span>
      </Stargazers>
    </Container>
  );
};
