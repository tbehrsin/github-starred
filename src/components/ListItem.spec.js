import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import ListItem from './ListItem';
import { ADD_STAR, REMOVE_STAR } from './StarButton';

const testRepository = {
  "url": "https://github.com/tbehrsin/github-starred",
  "nameWithOwner": "tbehrsin/github-starred",
  "name": "github-starred",
  "owner": {
    "avatarUrl": "https://avatars0.githubusercontent.com/u/45790596?v=4",
    "login": "github-starred"
  },
  "descriptionHTML": "<div>description html</div>",
  "stargazers": {
    "totalCount": 123
  },
  "id": "MDEwOlJlcG9zaXRvcnkxMjQyNTcyMzM=",
  "viewerHasStarred": false
};

const addStarMocks = [
  {
    request: {
      query: ADD_STAR,
      variables: { id: testRepository.id }
    },
    result: {
      data: {
        addStar: {
          clientMutationId: null,
          starrable: {
            viewerHasStarred: true
          }
        }
      }
    }
  }
];

const removeStarMocks = [
  {
    request: {
      query: REMOVE_STAR,
      variables: { id: testRepository.id }
    },
    result: {
      data: {
        removeStar: {
          clientMutationId: null,
          starrable: {
            viewerHasStarred: false
          }
        }
      }
    }
  }
];

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render((
    <MockedProvider>
      <ListItem repository={testRepository} />
    </MockedProvider>
  ), div);
  ReactDOM.unmountComponentAtNode(div);
});

it('stars a repository', async () => {
  testRepository.viewerHasStarred = false;
  const onChange = jest.fn();

  const component = renderer.create(
    <MockedProvider mocks={addStarMocks} addTypename={false}>
      <ListItem repository={testRepository} onChange={onChange} />
    </MockedProvider>
  );

  let tree = component.toJSON();
  expect(tree.children).toMatchSnapshot();

  const button = component.root.findByType('button');
  button.props.onClick();

  expect(onChange).not.toHaveBeenCalled();

  await wait(0);

  expect(onChange).toHaveBeenCalled();

  tree = component.toJSON();
  expect(tree.children).toMatchSnapshot();
});

it('unstars a repository', async () => {
  testRepository.viewerHasStarred = true;
  const onChange = jest.fn();

  const component = renderer.create(
    <MockedProvider mocks={removeStarMocks} addTypename={false}>
      <ListItem repository={testRepository} onChange={onChange} />
    </MockedProvider>
  );

  let tree = component.toJSON();
  expect(tree.children).toMatchSnapshot();

  const button = component.root.findByType('button');
  button.props.onClick();

  expect(onChange).not.toHaveBeenCalled();

  await wait(0);

  expect(onChange).toHaveBeenCalled();

  tree = component.toJSON();
  expect(tree.children).toMatchSnapshot();
});
