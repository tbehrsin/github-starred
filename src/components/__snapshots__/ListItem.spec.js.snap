// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stars a repository 1`] = `
Array [
  <div
    className="sc-htpNat fVdVsl"
  >
    <a
      className="sc-bxivhb jPCqqG"
      href="https://github.com/tbehrsin/github-starred"
      target="_new"
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 12 16"
        width={16}
      >
        <path
          d="M4 9H3V8h1v1zm0-3H3v1h1V6zm0-2H3v1h1V4zm0-2H3v1h1V2zm8-1v12c0 .55-.45 1-1 1H6v2l-1.5-1.5L3 16v-2H1c-.55 0-1-.45-1-1V1c0-.55.45-1 1-1h10c.55 0 1 .45 1 1zm-1 10H1v2h2v-1h3v1h5v-2zm0-10H2v9h9V1z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        github-starred
        /
        github-starred
      </span>
    </a>
    <button
      className="sc-bdVaJa klrspG"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 14 16"
        width={16}
      >
        <path
          d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        Star
      </span>
    </button>
  </div>,
  <div
    className="sc-ifAKCX fjTenm"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div>description html</div>",
      }
    }
  />,
  <div
    className="sc-EHOje eUBQSX"
  >
    <svg
      aria-hidden="true"
      className="StyledIconBase-sc-bdy9j4 jbPeIR"
      fill="currentColor"
      focusable="false"
      height={12}
      viewBox="0 0 14 16"
      width={12}
    >
      <path
        d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
        fillRule="evenodd"
      />
    </svg>
    <span>
      123
    </span>
  </div>,
]
`;

exports[`stars a repository 2`] = `
Array [
  <div
    className="sc-htpNat fVdVsl"
  >
    <a
      className="sc-bxivhb jPCqqG"
      href="https://github.com/tbehrsin/github-starred"
      target="_new"
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 12 16"
        width={16}
      >
        <path
          d="M4 9H3V8h1v1zm0-3H3v1h1V6zm0-2H3v1h1V4zm0-2H3v1h1V2zm8-1v12c0 .55-.45 1-1 1H6v2l-1.5-1.5L3 16v-2H1c-.55 0-1-.45-1-1V1c0-.55.45-1 1-1h10c.55 0 1 .45 1 1zm-1 10H1v2h2v-1h3v1h5v-2zm0-10H2v9h9V1z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        github-starred
        /
        github-starred
      </span>
    </a>
    <button
      className="sc-bdVaJa klrspG"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 14 16"
        width={16}
      >
        <path
          d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        Unstar
      </span>
    </button>
  </div>,
  <div
    className="sc-ifAKCX fjTenm"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div>description html</div>",
      }
    }
  />,
  <div
    className="sc-EHOje eUBQSX"
  >
    <svg
      aria-hidden="true"
      className="StyledIconBase-sc-bdy9j4 jbPeIR"
      fill="currentColor"
      focusable="false"
      height={12}
      viewBox="0 0 14 16"
      width={12}
    >
      <path
        d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
        fillRule="evenodd"
      />
    </svg>
    <span>
      123
    </span>
  </div>,
]
`;

exports[`unstars a repository 1`] = `
Array [
  <div
    className="sc-htpNat fVdVsl"
  >
    <a
      className="sc-bxivhb jPCqqG"
      href="https://github.com/tbehrsin/github-starred"
      target="_new"
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 12 16"
        width={16}
      >
        <path
          d="M4 9H3V8h1v1zm0-3H3v1h1V6zm0-2H3v1h1V4zm0-2H3v1h1V2zm8-1v12c0 .55-.45 1-1 1H6v2l-1.5-1.5L3 16v-2H1c-.55 0-1-.45-1-1V1c0-.55.45-1 1-1h10c.55 0 1 .45 1 1zm-1 10H1v2h2v-1h3v1h5v-2zm0-10H2v9h9V1z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        github-starred
        /
        github-starred
      </span>
    </a>
    <button
      className="sc-bdVaJa klrspG"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 14 16"
        width={16}
      >
        <path
          d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        Unstar
      </span>
    </button>
  </div>,
  <div
    className="sc-ifAKCX fjTenm"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div>description html</div>",
      }
    }
  />,
  <div
    className="sc-EHOje eUBQSX"
  >
    <svg
      aria-hidden="true"
      className="StyledIconBase-sc-bdy9j4 jbPeIR"
      fill="currentColor"
      focusable="false"
      height={12}
      viewBox="0 0 14 16"
      width={12}
    >
      <path
        d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
        fillRule="evenodd"
      />
    </svg>
    <span>
      123
    </span>
  </div>,
]
`;

exports[`unstars a repository 2`] = `
Array [
  <div
    className="sc-htpNat fVdVsl"
  >
    <a
      className="sc-bxivhb jPCqqG"
      href="https://github.com/tbehrsin/github-starred"
      target="_new"
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 12 16"
        width={16}
      >
        <path
          d="M4 9H3V8h1v1zm0-3H3v1h1V6zm0-2H3v1h1V4zm0-2H3v1h1V2zm8-1v12c0 .55-.45 1-1 1H6v2l-1.5-1.5L3 16v-2H1c-.55 0-1-.45-1-1V1c0-.55.45-1 1-1h10c.55 0 1 .45 1 1zm-1 10H1v2h2v-1h3v1h5v-2zm0-10H2v9h9V1z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        github-starred
        /
        github-starred
      </span>
    </a>
    <button
      className="sc-bdVaJa klrspG"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR"
        fill="currentColor"
        focusable="false"
        height={16}
        viewBox="0 0 14 16"
        width={16}
      >
        <path
          d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
          fillRule="evenodd"
        />
      </svg>
      <span>
        Star
      </span>
    </button>
  </div>,
  <div
    className="sc-ifAKCX fjTenm"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div>description html</div>",
      }
    }
  />,
  <div
    className="sc-EHOje eUBQSX"
  >
    <svg
      aria-hidden="true"
      className="StyledIconBase-sc-bdy9j4 jbPeIR"
      fill="currentColor"
      focusable="false"
      height={12}
      viewBox="0 0 14 16"
      width={12}
    >
      <path
        d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
        fillRule="evenodd"
      />
    </svg>
    <span>
      123
    </span>
  </div>,
]
`;
